client/src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnel/OutlineVpn.swift

```diff
@@ -89,11 +89,50 @@ public class OutlineVpn: NSObject {
       DDLogDebug("Calling NETunnelProviderSession.startTunnel([:])")
       try session.startTunnel(options: [:])
       DDLogDebug("NETunnelProviderSession.startTunnel() returned")
-      return ErrorCode.noError
     } catch let error as NSError  {
       DDLogError("Failed to start VPN: \(error.localizedDescription)")
       return ErrorCode.vpnStartFailure
     }
+
+    // Wait for start to be completed.
+    class TokenHolder {
+      var token: NSObjectProtocol?
+    }
+    let tokenHolder = TokenHolder()
+    await withCheckedContinuation { continuation in
+      tokenHolder.token = NotificationCenter.default.addObserver(forName: .NEVPNStatusDidChange, object: manager.connection, queue: nil) { notification in
+        let status = manager.connection.status
+        if status == .connected || status == .disconnected || status == .invalid {
+          DDLogDebug("Tunnel start done.")
+          if let token = tokenHolder.token {
+            NotificationCenter.default.removeObserver(token, name: .NEVPNStatusDidChange, object: manager.connection)
```

Comment:
Should you remove the observer if `status` is something other than the 3 you catch? Otherwise there'd be additional observers? Or does that already get cleaned up?

Reply:
I add only one observer, and remove it when we don't need it anymore. The observer will be notified of multiple events, which we need to ignore.

I added comments to clarify. I also reordered things to make it more robust: we now register the observer first, before we start the tunnel, to not run into a race condition.